--- a/src/sysoptions.h
+++ b/src/sysoptions.h
@@ -183,7 +183,9 @@ defined(__has_feature)
 #define LTC_ECC521
 #endif

-#define DROPBEAR_LTC_PRNG (DROPBEAR_ECC)
+#ifndef DROPBEAR_LTC_PRNG
+#define DROPBEAR_LTC_PRNG (DROPBEAR_ECC)
+#endif

 /* RSA can be vulnerable to timing attacks which use the time required for
  * signing to guess the private key. Blinding avoids this attack, though makes
@@ -436,7 +438,9 @@ defined(__has_feature)
 #define DROPBEAR_CLIENT_TCP_FAST_OPEN 0
 #endif

-#define DROPBEAR_TRACKING_MALLOC (DROPBEAR_FUZZ)
+#ifndef DROPBEAR_TRACKING_MALLOC
+#define DROPBEAR_TRACKING_MALLOC (DROPBEAR_FUZZ)
+#endif

 /* Used to work around Memory Sanitizer false positives */
 #if defined(__has_feature)
