--- a/Modules/timemodule.c
+++ b/Modules/timemodule.c
@@ -1800,15 +1800,15 @@ static int
     static const time_t YEAR = (365 * 24 + 6) * 3600;
     time_t t;
     struct tm p;
-    time_t janzone_t, julyzone_t;
+    long long janzone_t, julyzone_t;
     char janname[10], julyname[10];
     t = (time((time_t *)0) / YEAR) * YEAR;
     _PyTime_localtime(t, &p);
     get_zone(janname, 9, &p);
-    janzone_t = -get_gmtoff(t, &p);
+    janzone_t = -(long long)get_gmtoff(t, &p);
     janname[9] = '\0';
     t += YEAR/2;
     _PyTime_localtime(t, &p);
     get_zone(julyname, 9, &p);
-    julyzone_t = -get_gmtoff(t, &p);
+    julyzone_t = -(long long)get_gmtoff(t, &p);
     julyname[9] = '\0';
